### ISSUE TITLE
External pillars calling pillar_compile

### ISSUE THREAD

**Issue by gracinet** [saltstack/salt#21342]
This is a refactor wish / bunch of questions.

While working on #21268 with @ptonelli, we noticed that external pillars that work by bringing files to compile as if they were on the file system in the regular hierarchy (such as git_pillar and, precisely hg_pillar) were implemented by re-instantiation of the `Pillar` class with a changed `pillar_root`, and calling its `compile_pillar()`. 

The issue I have with this is that this apparently calls back all external pillars. 

Example for git_pillar:

```
    # Don't recurse forever-- the Pillar object will re-call the ext_pillar
    # function
    if __opts__['pillar_roots'].get(branch, []) == [pillar_dir]:
        return {}

    gitpil.update()  # this here is just the git fetch/pull (comment added to extract)

    opts = deepcopy(__opts__)

    opts['pillar_roots'][environment] = [pillar_dir]

    pil = Pillar(opts, __grains__, minion_id, branch)

    return pil.compile_pillar()
```

This isn't a clean execution flow. 

External pillar implementations such as git_pillar and friends should not need to call back the whole stack (it's possible that it was needed in the past and not nowadays, I don't have a full picture yet in my mind).

I wondering what happens if several such external pillars are present at the same time. Is there a possibility of git_pillar and svn_pillar triggering together an infinite loop ?

Besides, I noticed that the `Pillar` class has a special hardcoded case for git (hence reversing the logical dependencies of components):

```
        else:
            if key == 'git':
                ext = self.ext_pillars[key](self.opts['id'],
                                            val,
                                            pillar_dirs)
            else:
                ext = self.ext_pillars[key](self.opts['id'],
                                            pillar,
                                            val)
        return ext
```

**Comment by gracinet** [comment]
Forgot to say : we actually witnessed that the external pillar is called twice, hence that the protection against loop (commented as "don't recurse forever") is really necessary.

All this seen on the current develop branch

**Comment by jfindlay** [comment]
@gracinet, thanks for the excellent code review.  Have you talked with @terminalmage yet?  He seems to be the best expert in these areas that we have in house.

**Comment by jfindlay** [comment]
@gracinet, so, I've talked with @terminalmage, who has indicated that he's not the most expert on external pillars, but has done much excellent work on git related code.  Please feel free to submit pull requests for the things you've mentioned, thanks.

**Comment by gracinet** [comment]
@jfindlay @terminalmage thanks, I'll see what I can do in the coming week. I can use git_pillar as a working example, yes. We have tests ready for hg_pillar than I should be able to readily adapt to git, maybe I'll make a separate PR for that.

**Comment by gracinet** [comment]
git_pillar does not suffer from infinite recursion, because of the check on `pillar_dirs`, but that's somewhat specific to it, since it's the only one taking that arguments (and that's the special case in the calling code I was mentioning).
I could trigger an infinite one easily on the yet te be included hg_pillar (#21268)

**Comment by ekristen** [comment]
I am seeing this, it's causing my pillar system to not work properly, at least not `pillar.items` it seems that for state usage the correct pillar data is made available.

The behavior I'm seeing is that ext_pillar is being called twice. I'm using the new git pillar syntax and using __env__ and what is happening is the first time the correct branch is being resolved to the right environment, but then it gets called a second time and it goes to the base environment and master branch which for me is empty and returns an empty object to the minion. 

If I hack the pillar code and set `self.branch` to my environment, the pillar.items works as expected.

**Comment by terminalmage** [comment]
@ekristen I am almost 100% certain that this is not related. When git_pillar instantiates a ``Pillar`` class, it passes ``ext=False`` which prevents infinite recursion.

If this is really happening, there is a different cause.

**Comment by terminalmage** [comment]
I took a look at our external pillar modules and noticed that the s3 and svn ext_pillars were not passing ``ext=False``, so I opened https://github.com/saltstack/salt/pull/39232 to do this.

@gracinet appears to have fixed hg_pillar shortly after his most recent comment. I fixed git_pillar in b80da6e23a07bfacb15b3a873030d0f3b46089e2 (legacy) and 86d4fe2aa6b197e1cfd7e5d91820988284a19673 (rewritten git_pillar for 2015.8.0). The only other ext_pillars that could cause this recursion were s3 and svn, which I fixed just now in https://github.com/saltstack/salt/pull/39232, so I am going to close this issue.

@ekristen's I still maintain that your issue is different from this one. I'm not sure if the clarification I was able to give you earlier today in #39192 helped clear this up (as I noticed your previous comment here was from several days ago). If you are still experiencing multiple ext_pillar compilations, please open a new issue with a replicable use case, and mention me in the issue so that I get a notification.
